### DISCUSSION TITLE
Can you make a postgres function public (allowing unauthenticated users to hit it)

### DISCUSSION THREAD

**Discussion by carinahealth-alden** [supabase/supabase#21062] [Questions]
During onboarding I want to have a user enter an access code, since we're still in beta release. I want to check that access code against a table in my database. I don't want to make the whole table public, because it has important secure data in it.

I wrote a basic function that basically takes the access code, then checks against the DB to find if it exists, then returns true or false. When I plug it into the sign up workflow (I'm using weweb https://www.weweb.io/), it's not working - it seems to be requiring the user to be authenticated.

I found on the internet ([supabase help docs](https://supabase.com/docs/guides/database/functions#function-privileges)) a query to change the grants - but that didn't do anything. Also if this wasn't something I should have done, please let me know how to revert it :)

```
ALTER DEFAULT PRIVILEGES REVOKE EXECUTE ON FUNCTIONS FROM PUBLIC;
GRANT EXECUTE ON FUNCTION validate_login_code TO anon;
```

Any ideas? Open to any other ways to securely check if the code is valid

**Comment by mwoss** [comment]
As far as I know by default functions can be executed by any role, anon also. If you want to allow users to bypass security policies (not sure what's your current setup) you could change the function security level to `DEFINER`. You can read more about it here - [Database Functions
](https://supabase.com/docs/guides/database/functions#security-definer-vs-invoker)

**Comment by carinahealth-alden** [comment]
One of my guesses is that in this case the users are pre-sign up so they dont have a "user" yet - so there aren't any roles to associate the person to ... am I thinking about that right?

From the client I can successfully execute the function once the user is logged in, but still not able to do it if the user isn't logged in...

**Reply by mwoss** [reply]
If there's no user, Supabase assumes you are making a request as an anonymous user. I've tested it against my personal project and it works as expected. The lack of the `Authorization` header means that the call is done by anon user.
```sh
curl -X POST 'https://PROJECT_ID.supabase.co/rest/v1/rpc/PROCEDURE' \
-H "Content-Type: application/json" \
-H "apikey: PUBLIC_API_KEY"
```

**Reply by carinahealth-alden** [reply]
And you get the function to send the correct answer back to you? How did you update the function permissions to allow anonymous users to query it?

**Comment by GaryAustin1** [comment]
If you have RLS on the table to prevent it from being read by anon, then you have to make your function `security definer` type so that it can bypass the RLS on the table.  Even though anon might execute a function, by default setting of security INVOKER it has to meet RLS on all tables it accesses.

**Reply by carinahealth-alden** [reply]
Do you see security risks making it security definer? Any ways to mitigate that risk?

**Reply by GaryAustin1** [reply]
You are the one that wants to let anyone run it.  If you want to bypass RLS on your table to only read one row it is the only way to do it with a function.  Your risk is only that anyone can read the table one row at a time though.

The recommendation is to set the search path properly here:  https://supabase.com/docs/guides/database/functions#security-definer-vs-invoker

**Reply by carinahealth-alden** [reply]
I don’t WANT them to do it but I NEED to verify the code they provide before allowing them to register. Open to any other ways to do that :)

will read up on search paths

**Reply by GaryAustin1** [reply]
No matter what you have anon users with access to at least a row.  No way around that if that is what you need.
Security definer is fine just set the search path as shown is a safe thing to do.  If you are just selecting from the table you will be fine as that is all they can do with the function.

**Reply by carinahealth-alden** [reply]
I can also try to figure out an invite only flow  (triggered client side) that would avoid the public codes, just maybe above my technical grade…

**Reply by carinahealth-alden** [reply]
Ok yeah that makes sense

**Reply by carinahealth-alden** [reply]
Maybe I can find a way to have the backend issue a new user invite and I can call that by an authenticated user - checking out threads… I see you were on one related one earlier: https://github.com/supabase/supabase/issues/7928
